Cache support.

ZehMatt opened the pull request:
This PR implements caching to avoid sending the server the same code over and over, this is extremely noticeable with larger scripts. This is done by telling the server about each file first including the hash, if the server has not yet seen the file it will request it from the client otherwise it can use its cached code. Before it was nearly impossible to sequentially spawn the Starfall Processor with large scripts.

This PR also fixes a bug that would skip 1 byte of data when larger than 60000.

Before:
https://share.epic-domain.com/2019-01-16_23-29-18.mp4

After:
https://share.epic-domain.com/2019-01-17_01-57-46.mp4

Also yes this does not fix the huge lag when spawning its to tackle the the networking overhead, I might look into the lag next.

ZehMatt committed:
Implement basic hashing.

ZehMatt committed:
Implement cache support for processor scripts.

awilliamson commented on lua/starfall/sflib.lua:
```diff
@@ -701,48 +675,230 @@ if SERVER then
+		for ply, cachedFiles in pairs(codecache) do
+			for fname, entries in pairs(cachedFiles) do
+				for k, entry in pairs(entries) do
+					if CurTime() - entry.time >= (10 * 60) then
```
Perhaps make the 'minutes to cache' a SF variable so server administrators can tweak this? Might be relevant for high-volume servers.

awilliamson commented on lua/starfall/editor.lua:
```diff
@@ -1410,6 +1410,14 @@ if CLIENT then
 	end
 
+	-- FIXME: Compute a better hash to avoid collisions, only used for implementation completeness.
+	local function computeCodeHash(code)
```
Maybe we shouldn't worry too much about collisions; especially if it's for local caching? Chances of this are extremely low, and the result would be just sending some more data to the server again? Although the issue of using mis-matched code that has been cached against that hash could be an issue.

awilliamson commented on lua/starfall/sflib.lua:
```diff
@@ -701,48 +675,230 @@ if SERVER then
+	end
+
+	local function addCodeCache(ply, filename, hash, codeSize)
+		if codecache[ply][filename] == nil then
```
if `ply` returns nil, surely attempting to index `filename` upon this will result in an error `index a nil value`?

ZehMatt replied:
I don't believe ply can be nil when you are inside net.Receive, I might be wrong but I'll can catch that case.

awilliamson commented on lua/starfall/sflib.lua:
```diff
@@ -701,48 +675,230 @@ if SERVER then
+
+		local missing = {}
+		local mainfile = net.ReadString()
+		local count = net.ReadUInt(16)
```
count might want replacing with a more descriptive word. Like num_of_files, seeing as the net interactions aren't exactly documented..

awilliamson commented on lua/starfall/sflib.lua:
```diff
@@ -562,7 +536,7 @@ if SERVER then
 
         net.Start( "starfall_addnotify" )
-	net.WriteString( msg )
+		net.WriteString( msg )
```
This, and L556 seem opposites. Either have indenting on for writes after a start, or off. I'm leaning towards off as other writes in context are unindented.

awilliamson commented on lua/starfall/sflib.lua:
```diff
@@ -478,11 +478,15 @@ end
 
 if SERVER then
 	util.AddNetworkString( "starfall_requpload" )
+	util.AddNetworkString( "starfall_uploadlist")
```
Be careful of codestyle here. Few other spots in this file which slipped through. I would have made a suggestion, but GitHub review lacks the ability to do multiple line suggestions.

awilliamson commented on lua/starfall/editor.lua:
```diff
@@ -1410,6 +1410,14 @@ if CLIENT then
 		js( "editor.setFontSize(" .. GetConVarNumber( "sf_editor_fontsize" ) .. ")" )
 	end
 
+	-- FIXME: Compute a better hash to avoid collisions, only used for implementation completeness.
```
Careful of Codestyle in this file too.

awilliamson commented on lua/starfall/sflib.lua:
```diff
@@ -701,48 +675,230 @@ if SERVER then
+			local codeData = currentlist.files[v.filename]
+			local offset = 1
+			repeat
+
```
Lots of whitespace here.

awilliamson commented on lua/starfall/sflib.lua:
```diff
@@ -701,48 +675,230 @@ if SERVER then
+				net.SendToServer()
+
+				offset = offset + #data
+			until offset > #codeData
```
Do we want some form of timeout on sending? (future feature discussion)

awilliamson commented on lua/starfall/sflib.lua:
```diff
@@ -701,48 +675,230 @@ if SERVER then
+				net.WriteString( v.hash )
 			net.SendToServer()
-			if list then
-				SF.AddNotify( LocalPlayer(), list, NOTIFY_ERROR, 7, NOTIFYSOUND_ERROR1 )
```
Concerned about this being removed? Do players still get notified?

ZehMatt replied:
I wasn't exactly sure what this line does it did seem like a mistake of error handling but I guess its not supposed for errors?

Jazzelhawk replied:
The documentation for SF.Editor.BuildIncludesTable isn't clear but the function returns an error message as the second return value if it failed to build the includes table.

See here:
https://github.com/INPStarfall/Starfall/blob/develop/lua/starfall/editor.lua#L1502

awilliamson reviewed (changes requested):
Overall a very very good addition. Few places that codestyle needs to be addressed, and one or two places I've commented. Namely about the CRC collisions and the policy of dealing with them.

In your submission you highlight a before and after demo. Can I request one from a dedicated server (non-local), as I think some of the frame lag etc is caused by the server being local. Just to have a bit more coverage on behaviours.

ZehMatt commented:
I wasn't aware this Repo is actually alive and it was too late until I realized I used the wrong base. People told me to rather go for StarfallEx which seems to be maintained, would like to hear what you have to say about this.

awilliamson commented:
It has been a while since the last commit, that is true. It is more likely that StarfallEx is more prominent, due to the sheer volume of output (whether that speaks of quality, who knows).

The feature you have proposed is certainly desirable, and makes perfect sense. After @Xandaros brought it to my attention ( we somehow both missed the e-mail ) I asked "Why didn't we do this?!?". 
I haven't touched Lua in a while at this point, so I'm unsure on some of the code path execution issues I brought up, such as double indexing if the first index is `nil`.

Main philosophical question, if you will, is about how to handle CRCs and whether collisions matter in this context. The hashes are a nice fixed size output to identify larger code, as such there is no real security issue brought about by collisions, but a usability one.
Collisions will happen, there can be certain schemes to reduce their likelihood. Personally, I'd prefer to just mitigate any issues which do arise from collisions; they will 'eventually' happen. If we got a collision, the worst that happens is that new code runs as old cached code. If we're talking about a singular file CRC, chances of collision ( I believe ) is `1 in 2^32`. ( https://preshing.com/20110504/hash-collision-probabilities/ shows likelihood as a function of CRC n-bit, and k stored hashes.)

ZehMatt commented:
Fair points, I can resolve some of the reviewed code.

ZehMatt commented:
I have addressed all of the reviews. I should mention that the code has some conflicting variable names such as "file" or "list" which already exist. As for code consistency I noticed that a lot of code has whitespaces instead of tabs. I suggest using glualint and provide a glualint.json so people can easily Prettify the code with the given rules, but thats for another PR.

awilliamson commented:
I was not aware of GLuaLinter. I don't remember if we discussed linting. I know a few of the standard things like CI are hard to do with GLua and the environment stuff.

Are you able to show a video of it working in a dedicated server setting? And/or with multiple files? (main script + user library calls). No rush for it.

I would recommend you make a PR to StarfallEx, as I do think this is a useful feature, and more people might benefit from it in that repository too.

Xandaros commented:
Didn't know about GLuaLinter, either... and I was looking for something like that, too!

Thank you very much for mentioning it

ZehMatt commented:
I'm too lazy to setup a dedicated server I was hoping for more tests from the Team.

ZehMatt commented:
Gonna close this, its been lingering here for too long.
